server/consumer.go

```diff
@@ -456,16 +467,29 @@ func checkConsumerCfg(
 		}
 	}
 
-	// As best we can make sure the filtered subject is valid.
-	if config.FilterSubject != _EMPTY_ {
-		subjects := copyStrings(cfg.Subjects)
-		// explicitly skip validFilteredSubject when recovering
+	subjectFilters := subjectFilters(config.FilterSubject, config.FilterSubjects)
+
+	// Check subject filters overlap.
+	for outer, subject := range subjectFilters {
+		for inner, ssubject := range subjectFilters {
+			if inner != outer {
+				if subjectIsSubsetMatch(subject, ssubject) {
+					return NewJsConsumerOverlappingSubjectFiltersError()
+				}
+			}
+		}
+	}
+	if subjectFilters[0] != _EMPTY_ {
```

Comment:
Can subjectFilters be empty slice or nil?

We should check len(subjectFilters) > 0

Reply:
It can't.

```
func subjectFilters(filter string, filters []string) []string {
	if len(filters) == 0 {
		return []string{filter}
	}
	return filters
}
```

Will always return slice of len >= 1, but I don't like that part of the code too, as that's implicit and simple change in that function can break it.

What about changing  `subjectFilters` to:

```
// Utility for simpler if conditions in Consumer config checks.
// In future iteration, we can immediately create `o.subjf` and
// use it to validate things.
func subjectFilters(filter string, filters []string) ([]string, bool) {
	if len(filters) == 0 {
		return []string{filter}, filter == _EMPTY_
	}
	return filters, filters[0] == _EMPTY_
}
```

That simplifies most uses of it and removes the assumption that len() > 0.

Reply:
I am not a fan of that, if not filters subjf should be nil, and check for len(subjf) > 0 should always return true iff there are filters.

Suggest we rework that.

Reply:
What about..

```go
func subjectFilters(filter string, filters []string) ([]string, bool) {
	if filter != _EMPTY_ {
		filters = append(filters, filter)
	}
	return filters, len(filters) > 0
}
```

Reply:
Doing the rework now.